### ISSUE TITLE
Add a CustomCommandLineArgument interface

### ISSUE THREAD

**Issue by cmnbroad** [broadinstitute/barclay#130]
This would be useful when the doc and possible values for an argument are not static, but derived programmatically from code. For example, an argument where the values are drawn from the cross product of two existing enums, or some subset thereof:

```
enum A { A1, A2 }
enum B { B1, B2, B3 }

@Argument
public List<String> crossProductArgs;
```

where the values of `crossProductArgs` are drawn from the set { "A1-B1", "A1-B2", "A2-B1", "A2-B2", "B1-A1", ... }, with hyphen as a separator.

This link is to a private repository, but for those who can navigate to it, a real life (and slightly more complicated) version of this came up in https://github.com/broadinstitute/picard-private/pull/1218.

**Comment by magicDGS** [comment]
Would this enable also to programmatically to restrict enums? For example, with an use-case from HTSJDK, remove support for `unknown` sort order to be set if the argument is a `SortOrder` enum?

**Comment by cmnbroad** [comment]
Yes, perhaps custom validation or parsing could be another use for this.

**Comment by cmnbroad** [comment]
The CommandLineProgram base class should include methods that handle some of the warning messages as implemented in https://github.com/broadinstitute/gatk/pull/4429. Barclay should also include the TerminalColor some variant of class mentioned in https://github.com/broadinstitute/gatk/pull/4429#issuecomment-367141169.

**Comment by magicDGS** [comment]
I think that your previous comment is misplaced (it should be in https://github.com/broadinstitute/barclay/issues/127 or https://github.com/broadinstitute/barclay/pull/96)

**Comment by cmnbroad** [comment]
Yeah. I'll add it there.
